containers/daap-python-base/src/var/task/data_product_metadata.py

```diff
@@ -1,116 +1,326 @@
+            self.valid = True
+            if self.type.value == "metadata":
+                self.data = data_to_validate
+            else:
+                self.data_pre_convert = data_to_validate
         return self
 
-    def write_json_to_s3(self) -> None:
-        if hasattr(self, "data_product_metadata") and self.valid_metadata:
-            json_metadata = json.dumps(self.data_product_metadata)
+    def write_json_to_s3(self, write_key: str) -> None:
+        """
+        writes validated metadata or schema json files to s3
+        """
+
+        if hasattr(self, "data") and self.valid:
+            json_file = json.dumps(self.data)
             s3_client.put_object(
-                Body=json_metadata,
-                Bucket=self.metadata_bucket,
-                Key=self.metadata_key,
+                Body=json_file,
+                Bucket=self.write_bucket,
+                Key=write_key,
                 **{
                     "ACL": "bucket-owner-full-control",
                     "ServerSideEncryption": "AES256",
                 },
             )
-            self.logger.info("Data Product metadata written to s3")
-        else:
-            self.logger.error(
-                "Metadata need to be validated before writing to s3, run the validate() class method"
+            self.logger.info(
+                f"Data Product {self.type} written to s3 at {self.write_bucket}/{write_key}"
             )
+        else:
+            self.logger.error(f"{self.type} is not valid to write.")
             raise ValidationError(
                 "Metadata must be validated before being written to s3."
             )
+
+    def load(self):
+        """loads the latest version of json file saved in s3"""
+        if self.exists:
+            self.latest_version_saved_data = read_json_from_s3(
+                f"s3://{self.write_bucket}/{self.latest_version_key}"
+            )
+        else:
+            self.logger.error("There is no metadata to load")
+        return self
+
+
+class DataProductSchema(BaseJsonSchema):
+    """
+    class to handle creation and updating of
+    schema json files relating to data product tables
+    """
+
+    def __init__(
+        self,
+        data_product_name: str,
+        table_name: str,
+        logger: DataPlatformLogger,
+        input_data: dict | None,
+    ):
+        # returns path of latest schema or v1 if it doesn't exist
+        bucket_path = DataProductConfig(name=data_product_name).schema_path(table_name)
+
+        super().__init__(
+            data_product_name=data_product_name,
+            logger=logger,
+            json_type=JsonSchemaName("schema"),
+            bucket_path=bucket_path,
+            input_data=input_data,
+            table_name=table_name,
+        )
+
+        if not self._does_data_product_metadata_exist():
+            self.logger.error("Data product metadata not yet registered.")
+            self.has_registered_data_product = False
```

Comment:
I think you could drop this boolean and instead set `parent_data_product_metadata` to `None` if it's missing

Reply:
Thanks, going to keep as is at the moment as it interacts with the way i've coded the lambda handler